### ISSUE TITLE
Support calling AWS generate-db-auth-token to get DB password for AWS RDS/Aurora Postgres

### ISSUE THREAD

**Issue by suntala** [grafana/grafana#75965]
### Discussed in https://github.com/grafana/grafana/discussions/48170

<div type='discussions-op-text'>

<sup>Originally posted by **krestivo-kdinfotech** May 13, 2020</sup>
**What would you like to be added**:

Support calling AWS generate-db-auth-token API to obtain a Postgres password before opening or re-opening any DB connection.

**Why is this needed**:
When running in AWS with IAM auth, you don' t have a password, you use a temporary token which goes into the PGPASSWORD field of the Postgres DB connection. This token times out after 15 minutes.

Grafana seems to close and open DB connections often, even with pooling, rather than keeping a single connection open. When this connection goes stale, Grafana seems to attempt to reconnect to the database. However, the token is no longer any good and you get RDS password errors.

The fix would be for Grafana to support a hook that could be run prior to it attempting to connect to a database or reconnect to a database, which the user could specify an external (shell) function to obtain the correct RDS token.</div>

**Comment by suntala** [comment]
Investigate further. It may need to be split into two issues, one for Backend Platform and whichever team owns SQL data sources.

**Comment by keymon** [comment]
Without looking much at the code of grafana, one problem is that lib/pq does not support callbacks for refreshing the creds.

but I have a potential workaround for this issue, by leveraging PGPASSFILE.  the idea is to point pgpassfile=... to a tmp file with the credentials. Then start in the background a go-routine that keeps updating the temporary file with the password provided by AWS RDS IAM.

I have here a snippet/function to do this, which we used in some project https://gist.github.com/keymon/e9b91d3644254bbc8dc587ca57d9eece . The snippet sets PGPASSFILE env var, which works, but now now it can even passed in the URL parameter to the DB, so we avoid polute the environment vars https://github.com/lib/pq/pull/1129

**Comment by nishkrishnan** [comment]
You'd have to implement the connector interface: https://pkg.go.dev/database/sql/driver#Connector. In that you can generate the token and open a new postgres connection using that.

The custom connector would be used here https://pkg.go.dev/database/sql#OpenDB.  We've done this internally for our postgres services.

**Comment by fatbasstard** [comment]
Looking for the same feature for MySQL. We've got a new database we'd like to connect to with IAM Authentication enabled, so that feature would be nice to have

**Comment by jarpoole** [comment]
I need the same feature for Azure managed postgres and Active Directory authentication as well...

**Comment by misterek** [comment]
Another vote here for RDS/Aurora.  Supporting this allows us to handle credentials more effectively.

**Comment by theintz** [comment]
We are switching our RDS postgres instances to IAM auth. Would love to see this implemented so that we can get rid of the static credentials.

**Comment by SJM-J2** [comment]
Tossing my own hat in the ring...supporting IAM auth would be a huge help...currently this DB is the odd man out in our infra

**Comment by dlightman** [comment]
Grafana is one of the last components in our stack to not support IAM authentication to RDS. I too would love to see this feature added.

**Comment by keymon** [comment]
You can try the pgpassfile approach I mentioned above: https://github.com/grafana/grafana/issues/75965#issuecomment-2549528832 ? I didn't try it myself, but it shall work. I did this for other projects.

Basically, for grafana:

 1. Add a sidecar (if in k8s) or a process running in parallel updating a pgfile. Like this https://gist.github.com/keymon/e9b91d3644254bbc8dc587ca57d9eece
 2. pass PGPASSFILE to use that file

Alternatively you can make that snippet I paste a command line tool that runs grafana, as a wrapper.

I also did a PR to lib/pq to accept pgpassfile in the connection url, but they didn't release any version since may 2023 and I doubt it is in grafana 🤷

**Comment by taraspos** [comment]
For datasources this is available as an enterprise plugin now:
- https://grafana.com/grafana/plugins/grafana-aurora-datasource/
- https://grafana.com/whats-new/2024-08-22-aurora-data-source-for-grafana/

**Comment by Fokines** [comment]
does it available to connect to a database via IAM? not as a datasource

**Comment by matt-veitas-eb** [comment]
One solution for AWS RDS might to be leverage the AWS RDS proxy (https://aws.amazon.com/rds/proxy/)

1. Grafana → Proxy (simple password)
2. Proxy → RDS (IAM tokens, auto-refreshed)
3. Zero changes to Grafana

**Comment by sigma67** [comment]
Are contributions welcome on this topic?
